### PR TITLE
Add support customized IPSans and DNSSans for server-tls server cert for vault

### PR DESCRIPTION
**Pull request by jmurret**
Changes proposed in this PR:
- Add support customized IPSans and DNSSans for server-tls server cert for vault

How I've tested this PR:
- [x] unit tests
- [ ] manual verification in a consul/vault/k8s environment

How I expect reviewers to test this PR:
manual verification in a consul/vault/k8s environment

Checklist:
- [x] Tests added
- [x] CHANGELOG entry added

### CODE REVIEW

**Review by kschoche — APPROVED** [review]
I think this looks great, assuming that you get the manual test working and confirmed I think it's good to merge after.
One comment on my way :
Is there any way you could roll the new templating logic into a helper function? It is really a lot of text to all inline. Thoughts?

File: charts/consul/templates/_helpers.tpl [hunk 1]
```diff
@@ -32,15 +32,15 @@ as well as the global.name setting.
...
             {{ "{{" }}- .Data.certificate -{{ "}}" }}
             {{ "{{" }}- end -{{ "}}" }}
 {{- end -}}
 
 {{- define "consul.serverTLSKeyTemplate" -}}
  |
             {{ "{{" }}- with secret "{{ .Values.server.serverCert.secretName }}" "{{ printf "common_name=server.%s.%s" .Values.global.datacenter .Values.global.domain }}"
-            "ttl=1h" "alt_names={{ include "consul.serverTLSAltNames" . }}" "ip_sans=127.0.0.1" -{{ "}}" }}
+            "ttl=1h" "alt_names={{ include "consul.serverTLSAltNames" . }}{{- if .Values.server.tls -}}{{- if .Values.server.tls.serverAdditionalDNSSANs -}}{{- range $san := .Values.server.tls.serverAdditionalDNSSANs }},{{ $san }} {{- end -}}{{- end -}}{{- end -}}" "ip_sans=127.0.0.1{{- if .Values.server.tls -}}{{- if .Values.server.tls.serverAdditionalIPSANs -}}{{- range $ipsan := .Values.server.tls.serverAdditionalIPSANs }},{{ $ipsan }} {{- end -}}{{- end -}}{{- end -}}" -{{ "}}" }}
```
Comment (ishustava) [inline charts/consul/templates/_helpers.tpl:43]: drive-by comment: these values should come from `.Values.global.tls` not `.Values.server.tls`. The goal I think is to re-use existing values: https://github.com/hashicorp/consul-k8s/blob/216fbdad145912a3e0e03b9633af4ee6fbb24ffd/charts/consul/values.yaml#L282

File: charts/consul/test/unit/server-statefulset.bats [hunk 2]
```diff
@@ -1786,6 +1786,66 @@ load _helpers
...
+    --set 'server.tls.serverAdditionalDNSSANs[0]=*.foo.com' \
+    --set 'server.tls.serverAdditionalDNSSANs[1]=*.bar.com' \
+    . | tee /dev/stderr |
+      yq -r '.spec.template' | tee /dev/stderr)
+
+  local actual=$(echo $object |
+      yq -r '.metadata.annotations["vault.hashicorp.com/agent-inject-template-servercert.crt"]' | tee /dev/stderr)
+  local expected=$'{{- with secret \"pki_int/issue/test\" \"common_name=server.dc2.consul\"\n\"ttl=1h\" \"alt_names=localhost,RELEASE-NAME-consul-server,*.RELEASE-NAME-consul-server,*.RELEASE-NAME-consul-server.default,*.RELEASE-NAME-consul-server.default.svc,*.server.dc2.consul,*.foo.com,*.bar.com\" \"ip_sans=127.0.0.1\" -}}\n{{- .Data.certificate -}}\n{{- end -}}'
+  echo "Expected: ${expected}"
+  echo "Actual: ${actual}"
```
Comment (ishustava) [inline charts/consul/test/unit/server-statefulset.bats:1811]: Did you mean to keep these here or was it for debugging?

Comment (jmurret) [reply charts/consul/test/unit/server-statefulset.bats:1811]: ooh.  thx!

File: charts/consul/templates/_helpers.tpl (hunk 1 above)
Comment (jmurret) [reply charts/consul/templates/_helpers.tpl:43]: :face-palm.  Thank you!

**Comment by jmurret** [comment]
@kschoche great call on the helper function!  will add

**Review by t-eckert — APPROVED** [review]
This looks excellent, John! Great work.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 IMPROVEMENTS:
 * Helm
   * Vault: Allow passing arbitrary annotations to the vault agent. [[GH-1015](https://github.com/hashicorp/consul-k8s/pull/1015)]
+  * Vault: Add support customized IPSans and DNSSans for server-tls server cert for vault [GH-1020](https://github.com/hashicorp/consul-k8s/pull/1020)
```
Comment (ishustava) [inline CHANGELOG.md:5]:
```suggestion
  * Vault: Add support for customized IP and DNS SANs for server cert in Vault. [[GH-1020](https://github.com/hashicorp/consul-k8s/pull/1020)]
```

File: charts/consul/templates/_helpers.tpl [hunk 4]
```diff
@@ -32,23 +32,31 @@ as well as the global.name setting.
...
+            "ttl=1h" "alt_names={{ include "consul.serverTLSAltNames" . }}" "ip_sans=127.0.0.1{{ include "consul.serverAdditionalIPSANs" . }}" -{{ "}}" }}
             {{ "{{" }}- .Data.private_key -{{ "}}" }}
             {{ "{{" }}- end -{{ "}}" }}
 {{- end -}}
 
 {{- define "consul.serverTLSAltNames" -}}
 {{- $name := include "consul.fullname" . -}}
 {{- $ns := .Release.Namespace -}}
-{{ printf "localhost,%s-server,*.%s-server,*.%s-server.%s,*.%s-server.%s.svc,*.server.%s.%s" $name $name $name $ns $name $ns (.Values.global.datacenter ) (.Values.global.domain) }}
+{{ printf "localhost,%s-server,*.%s-server,*.%s-server.%s,*.%s-server.%s.svc,*.server.%s.%s" $name $name $name $ns $name $ns (.Values.global.datacenter ) (.Values.global.domain) }}{{ include "consul.serverAdditionalDNSSANs" . }}
```
Comment (kschoche) [inline charts/consul/templates/_helpers.tpl:51]: This looks so much better, great work. I totally missed the typos that Iryna pointed out bc it was all one giant line of text and my brain wasnt up to the task. Much better!

**Review by ishustava — APPROVED** [review]
Looks great!!

**Comment by jmurret** [comment]
Screen Shot 2022-02-08 at 4 42 05 PM

Verified DNS SANs and IP SANs flowing through to servercert.  Going to merge.
